### PR TITLE
GEM ML examples

### PR DESCRIPTION
**Pull request by meengel**
As discussed we start the pull-request to further discuss and include the GEM ML examples up to their current stage into the official eo-learn-examples repository. They are tested on both Windows and Linux machines. The Deforestation Example has been presented within the [GEM Webinar](https://www.globalearthmonitor.eu/webinar) in February.

Parts of them will likely be changed, extracted or expanded. For example, the configuration and augmentation libraries will be published as own packages later on. Additionally, the linking of eo-learn and PyTorch, i.e. the PyTorchTasks, will be made available in an own package as well or in a new subpackage of eo-learn, namely eo-torch. The [TDigestTask has been merged into the develop-branch of eo-learn](https://github.com/sentinel-hub/eo-learn/pull/603) and has to be included after the official release as well. For now, all features, EOTasks or utils will be included as is to guarantee that some working version of the examples is ready by the end of the Global Earth Monitor project.

@batic @zigaLuksic Please have a look and let us know what has to be done!

### CODE REVIEW

File: GEM-ML/.gitignore
Comment (batic) [inline GEM-ML/.gitignore:1]: It would be simpler to merge all the .gitignore files into one single one, but I leave it up to you to decide.

File: GEM-ML/Example_Treecurrent/README.md
Comment (batic) [inline GEM-ML/Example_Treecurrent/README.md:1]: I'd add the link to ConvLSTM_pytorch in this readme for proper attribution (btw, it is released under MIT), skip the extra files in the GEM-ML/Example_Treecurrent/models/ConvLSTM_pytorch, and only keep convlstm.py in the ML/Example_Treecurrent/models.

File: GEM-ML/requirements.txt
Comment (batic) [inline GEM-ML/requirements.txt:1]: 🤯 

Are all these really needed? Perhaps you could try running something like [creosote](https://github.com/fredrikaverpil/creosote) to see if some are non needed?

**Review by batic — COMMENTED** [review]
First pass :) 

In any case, git(hub) is not good for reviewing notebooks, so expect some extra comments...

Best,
M

File: GEM-ML/.gitignore
Comment (ColinMoldenhauer) [reply GEM-ML/.gitignore:1]: Done

File: GEM-ML/Example_Treecurrent/README.md
Comment (ColinMoldenhauer) [reply GEM-ML/Example_Treecurrent/README.md:1]: Done

File: GEM-ML/requirements.txt
Comment (ColinMoldenhauer) [reply GEM-ML/requirements.txt:1]: Multiple tools to clean up the requirements file have failed on my machine, so I'll leave it for now. If it is absolutely crucial, I can take another look at this.

Comment (batic) [reply GEM-ML/requirements.txt:1]: No, I don't think it is crucial, but it might turn some users away...

File: GEM-ML/requirements.txt
Comment (batic) [inline GEM-ML/requirements.txt:142]:
```
pywin32==305
pywinpty==2.0.10
```

These two can only be installed on Windows, so the `pip install -r requirements.txt` failed for me on MacOS. Please separate the two into `requirements-win.txt` or similar (or remove altogether if not needed).

File: GEM-ML/README.md
Comment (batic) [inline GEM-ML/README.md:17]: I'd try to be more specific here, e.g. something along the lines:

 * `pip install -r requirements.txt`
 * install pytorch (+link to installation instructions for pytorch)
 * instructions for any other libraries not covered by requirements file
 * install GEM-ML by executing `pip install -e .` when your working directory is GEM-ML

File: GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb:1]: The notebook should provide instructions what additional packages should be installed, e.g. `segmentation_models_pytorch`.

In 01_DataAcquisition it transpires that `captum` should be installed, and `tensorboardX`. Probably there are more requirements that are missing in the `requirements.txt` file, reducing the usability of the examples. I'd suggest you try from empty environment and make sure that things work from the scratch.

File: GEM-ML/Example_DeforestationDetection/01_DataAcquisition.ipynb
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/01_DataAcquisition.ipynb:1]: I don't use conda, so `print("Environment:", os.environ['CONDA_DEFAULT_ENV'])` fails for me (in the first cell already).

File: GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.py
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.py:1]: This file can probably be removed.

File: GEM-ML/Example_Treecurrent/00_Configuration.ipynb
Comment (batic) [inline GEM-ML/Example_Treecurrent/00_Configuration.ipynb:1]: Checking the imports, only these are needed in this notebook:

```python

import datetime as dt
import os
import sys

import folium
import geopandas as gpd
import matplotlib.pyplot as plt
import numpy as np
from folium import plugins as foliumplugins
from libs.ConfigME import Config
from sentinelhub import DataCollection, UtmZoneSplitter
from shapely.geometry import Polygon
from tasks.PickIdxTask import PickIdxTask
from tasks.SaveValidTask import SaveValidTask
from tasks.TDigestTask import TDigestTask
from utils.parse_time_interval_observations import parse_time_interval_observations

from eolearn.core import EOExecutor, EONode, EOPatch, EOWorkflow, FeatureType, MapFeatureTask, OverwritePermission, \
    ZipFeatureTask
from eolearn.features.feature_manipulation import SpatialResizeTask
from eolearn.features.utils import ResizeLib, ResizeMethod
from eolearn.io import ImportFromTiffTask, SentinelHubInputTask
```

Please update.

Comment (batic) [reply GEM-ML/Example_Treecurrent/00_Configuration.ipynb:1]: Sorry, this is for 01_DataAcquisition notebook. 🙈

File: GEM-ML/Example_Treecurrent/01_DataAcquisition.ipynb
Comment (batic) [inline GEM-ML/Example_Treecurrent/01_DataAcquisition.ipynb:1]: Only the following imports are needed, please fix:

```python
import datetime as dt
import os
import sys

import folium
import geopandas as gpd
import matplotlib.pyplot as plt
import numpy as np
from folium import plugins as foliumplugins
from libs.ConfigME import Config
from sentinelhub import DataCollection, UtmZoneSplitter
from shapely.geometry import Polygon
from tasks.PickIdxTask import PickIdxTask
from tasks.SaveValidTask import SaveValidTask
from tasks.TDigestTask import TDigestTask
from utils.parse_time_interval_observations import parse_time_interval_observations

from eolearn.core import EOExecutor, EONode, EOPatch, EOWorkflow, FeatureType, MapFeatureTask, OverwritePermission, \
    ZipFeatureTask
from eolearn.features.feature_manipulation import SpatialResizeTask
from eolearn.features.utils import ResizeLib, ResizeMethod
from eolearn.io import ImportFromTiffTask, SentinelHubInputTask
```

File: GEM-ML/Example_Treecurrent/03_TrainingValidation.ipynb
Comment (batic) [inline GEM-ML/Example_Treecurrent/03_TrainingValidation.ipynb:1]: Only the following imports seem to be used in this notebook, please fix:

```python
import datetime as dt
import os
import platform
import sys

import matplotlib.pyplot as plt
import natsort
import numpy as np
import torch
from libs import AugmentME
from libs.ConfigME import Config, importME
from libs.Dataset_eolearn import Dataset_eolearn
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from tensorboard import notebook
from tensorboardX import SummaryWriter
from utils.transforms import Torchify

from eolearn.core import EOPatch, FeatureType
```

File: GEM-ML/Example_Treecurrent/02_DataNormalisation.ipynb
Comment (batic) [reply GEM-ML/Example_Treecurrent/02_DataNormalisation.ipynb:1]: Apologies, I've made a mess; it is really difficult to nicely review notebooks. These comments were meant for 01...04 notebooks in Deforestation detection.

File: GEM-ML/Example_DeforestationDetection/02_DataNormalization.ipynb
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/02_DataNormalization.ipynb:1]: The needed imports:

```python
import os
import platform
import sys
import time

import matplotlib.pyplot as plt
import numpy as np
from libs.ConfigME import Config
from libs.MergeTDigests import mergeTDigests
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from utils.transforms import Torchify

from eolearn.core import FeatureType
```

File: GEM-ML/Example_DeforestationDetection/03_TrainingValidationTesting.ipynb
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/03_TrainingValidationTesting.ipynb:1]: Cleaned imports:

```python
import datetime as dt
import os
import platform
import sys

import matplotlib.pyplot as plt
import natsort
import numpy as np
import torch
from libs import AugmentME
from libs.ConfigME import Config, importME
from libs.Dataset_eolearn import Dataset_eolearn
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from tensorboard import notebook
from tensorboardX import SummaryWriter
from utils.transforms import Torchify

from eolearn.core import EOPatch, FeatureType
```

File: GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.ipynb
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.ipynb:1]: Cleaned imports:

```python
import datetime as dt
import os
import sys
import time

import folium
import geopandas as gpd
import numpy as np
import torch
from folium import plugins as foliumplugins
from libs import AugmentME
from libs import ExecuteME
from libs.ConfigME import Config, importME
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from sentinelhub import DataCollection, UtmZoneSplitter
from shapely.geometry import Polygon
from tasks.PyTorchTasks import ModelForwardTask
from utils.transforms import predict

from eolearn.core import EONode, EOWorkflow, FeatureType, MapFeatureTask, OverwritePermission, SaveTask
from eolearn.io import ExportToTiffTask, SentinelHubInputTask
```

File: GEM-ML/Example_DeforestationDetection/README.md
Comment (batic) [inline GEM-ML/Example_DeforestationDetection/README.md:10]: Please add a direct link to download the `JRC_TMF_AnnualChange_v2_2021_SAM_ID30_N0_W60.tif` file.

File: GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb
Comment (batic) [reply GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb:1]: It would also make sense to add check for sh config, like in Treecurent example:

```python
if not config["SHconfig"].sh_client_id or not config["SHconfig"].sh_client_secret:
    print("Warning! To use Process API, please provide the credentials (OAuth client ID and client secret).")
```

File: GEM-ML/Example_WaterSegmentation/00_Configuration.ipynb
Comment (batic) [inline GEM-ML/Example_WaterSegmentation/00_Configuration.ipynb:1]: A typo (lookt):

> Let's load our credentials for Sentinel Hub from storage. If you should not have stored your credentials on disk yet, you should have a lookt at this [notebook](https://gitlab.lrz.de/mkoerner/projects-and-proposals/projects/2020_GEM/howto-eo-learn/-/blob/main/1_Configuration/tutorial1_config.ipynb).

Also, the notebook you are referring to is internal to TUM. Please fix.

File: GEM-ML/Example_WaterSegmentation/01_DataAcquisition.ipynb
Comment (batic) [reply GEM-ML/Example_WaterSegmentation/01_DataAcquisition.ipynb:1]: @meengel Are the "yet to come" notebook coming?

File: GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb
Comment (batic) [inline GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: Also here there are many unused imports. Please fix.

Comment (batic) [reply GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: @meengel all the notebooks in this folder have the same header info with "yet to come". Please add notebooks to come or cleanup the header descriptions.

File: GEM-ML/Example_WaterSegmentation/03_TrainingValidationTesting.ipynb
Comment (batic) [inline GEM-ML/Example_WaterSegmentation/03_TrainingValidationTesting.ipynb:1]: PyCharm only says the following imports are needed:

```python
import datetime as dt
import os
import platform
import sys

import matplotlib.pyplot as plt
import natsort
import numpy as np
import torch
from libs import AugmentME
from libs.ConfigME import Config, importME
from libs.Dataset_eolearn import Dataset_eolearn
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from tensorboard import notebook
from tensorboardX import SummaryWriter
from utils.transforms import Torchify

from eolearn.core import EOPatch, FeatureType
```

Please fix.

File: GEM-ML/Example_WaterSegmentation/04_PyTorchTasks_ModelForwardTask.ipynb
Comment (batic) [inline GEM-ML/Example_WaterSegmentation/04_PyTorchTasks_ModelForwardTask.ipynb:1]: PyCharm says only the following imports are needed in the notebook, please fix:

```python
import datetime as dt
import os
import sys
import time

import folium
import geopandas as gpd
import numpy as np
import rasterio
import torch
from folium import plugins as foliumplugins
from libs import AugmentME
from libs import ExecuteME
from libs.ConfigME import Config
from libs.QuantileScaler_eolearn import QuantileScaler_eolearn_tdigest
from sentinelhub import DataCollection, UtmZoneSplitter
from shapely.geometry import Polygon
from tasks.PyTorchTasks import ModelForwardTask
from utils.parse_time_interval_observations import parse_time_interval_observations
from utils.rasterio_reproject import rasterio_reproject
from utils.transforms import predict

from eolearn.core import EONode, EOWorkflow, FeatureType, OverwritePermission, SaveTask
from eolearn.io import ExportToTiffTask, SentinelHubInputTask
```

**Comment by batic** [comment]
I've now had more detailed, but still a quick look at the notebooks, and provided feedback. After you take the requests into account, we can proceed with merge. 

It's really good to see the examples you've put together!

File: GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/00_Configuration.ipynb:1]: - [ ] requirements (in progress)
- [x]  credential warning

File: GEM-ML/Example_DeforestationDetection/01_DataAcquisition.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/01_DataAcquisition.ipynb:1]: - [x] removed print of *CONDA_DEFAULT_ENV*
- [x] adapted `SpatialResizeTask` 
- [x] include missing file
- [x] imports (slightly adapted)

File: GEM-ML/Example_DeforestationDetection/02_DataNormalization.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/02_DataNormalization.ipynb:1]: done

File: GEM-ML/Example_DeforestationDetection/03_TrainingValidationTesting.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/03_TrainingValidationTesting.ipynb:1]: done, slightly adapted

File: GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.py
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.py:1]: done

File: GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_DeforestationDetection/04_Inference_Clouds.ipynb:1]: done

File: GEM-ML/Example_WaterSegmentation/00_Configuration.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/00_Configuration.ipynb:1]: done

File: GEM-ML/Example_WaterSegmentation/01_DataAcquisition.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/01_DataAcquisition.ipynb:1]: done

File: GEM-ML/Example_WaterSegmentation/03_TrainingValidationTesting.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/03_TrainingValidationTesting.ipynb:1]: done

File: GEM-ML/Example_WaterSegmentation/04_PyTorchTasks_ModelForwardTask.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/04_PyTorchTasks_ModelForwardTask.ipynb:1]: done

File: GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb
Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: done

Comment (batic) [reply GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: @ColinMoldenhauer I still see "yet to come", so I assume "done" is meant for fixing unused imports. What are your plans regarding missing notebooks?

Comment (batic) [reply GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: Also, don't forget to push updated files. The last change for this file I see is from Michael, from few months ago.

Comment (ColinMoldenhauer) [reply GEM-ML/Example_WaterSegmentation/02_DataNormalization.ipynb:1]: Sorry for the confusion, I commented as a personal reference to keep track of what has been done :) I will merge the changes into here after discussing a few things on Friday!

**Comment by batic** [comment]
@ColinMoldenhauer could you please let me know if you pushed all the changes and I can have a final look?

**Review by batic — APPROVED** [review]
Thanks! 🚀

**Pull request merged**
